cluster/discovery/discovery_mcast.go

```diff
@@ -0,0 +1,374 @@
+	defer b.Unlock()
+
+	_, exists := b.Nodes[ne.Id]
+	if !exists {
+		return nil, fmt.Errorf("Unable to find node %v in discovery db", ne.Id)
+	}
+	delete(b.Nodes, ne.Id)
+	b.Version++
+
+	// create a copy
+	ci := *(&b.ClusterInfo)
+
+	return &ci, nil
+}
+
+func (b *discoveryMcast) Enumerate() (*ClusterInfo, error) {
+	b.Lock()
+	ci := *(&b.ClusterInfo) // Make a copy
+	b.Unlock()
+
+	return &ci, nil
+}
+
+func (b *discoveryMcast) addAllAndNotifyWatch(stream io.ReadCloser) error {
+	var ci ClusterInfo
+	if body, err := ioutil.ReadAll(stream); err != nil {
+		dlog.WithField("err", err).Warnf("Could not read config-request")
+		return err
+	} else if err := json.Unmarshal(body, &ci); err != nil {
+		dlog.WithField("err", err).Warnf("Could not parse config-request")
+		return err
+	}
+
+	//// ideally TRACE message
+	//logrus.Debugf("addAllAndNotifyWatch() got %+v, have %+v", ci, b.ClusterInfo)
+
+	b.Lock()
+	defer b.Unlock()
+
+	stGotNew, stGotUpdates := false, false
+	for k, v := range ci.Nodes {
+		if _, exists := b.Nodes[k]; !exists {
+			dlog.Infof("Discovered node %v", v)
+			b.Nodes[k] = v
+			stGotNew = true
+		} else if ci.Version > b.Version {
+			dlog.Infof("Updated node %v", v)
+			b.Nodes[k] = v
+			stGotUpdates = true
+		} else {
+			dlog.Debugf("Ignoring update to %s (version too low)", k)
+		}
+	}
+
+	if ci.Version > b.Version {
+		b.Version = ci.Version
+	}
+
+	if stGotNew {
+		b.Size = len(b.Nodes)
+		b.Version++
+	}
+
+	// ideally TRACE message
+	dlog.Debugf("addAllAndNotifyWatch() -- post additions, my updated state is %v", b.ClusterInfo)
+
+	if b.wcb != nil && (stGotNew || stGotUpdates) {
+		ci := *(&b.ClusterInfo)
+		b.wcb(&ci, nil)
+	}
+	return nil
+}
+
+// restConfigExchange is a REST-processing function, which will process the passed config
+func (b *discoveryMcast) restConfigExchange(w http.ResponseWriter, r *http.Request) {
+	dlog.Debugf("Processing REST config update via %s %v", r.Method, r.RequestURI)
+
+	w.Header().Add("Server", "OpenLib-Discovery")
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		r.Body.Close()
+		return
+	}
+
+	// Process input config
+	if err := b.addAllAndNotifyWatch(r.Body); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(err.Error()))
+		r.Body.Close()
+		return
+	}
+
+	// Send back my updated config
+	ci, _ := b.Enumerate()
+	reply, err := json.Marshal(ci)
+	if err != nil {
+		dlog.WithField("err", err).Warnf("Could not encode my nodes")
+		return
+	}
+	_, err = w.Write(reply)
+	if err != nil {
+		dlog.WithField("err", err).Warnf("Error sending back my nodes")
+	} else {
+		dlog.Debugf("Sent back my nodes") // string(reply)?
+	}
+}
+
+func (b *discoveryMcast) mcastConfigExchange(udpAddr *net.UDPAddr) error {
+	ci, _ := b.Enumerate()
+	payload, err := json.Marshal(ci)
+	if err != nil {
+		dlog.WithField("err", err).Warnf("Failed serializing config %+v", ci)
+		return err
+	}
+
+	url := fmt.Sprintf("http://%s:%s", udpAddr.IP.String(), b.port)
```

Comment:
Can you put out comments on how this thing works ?
It is a bit confusing as to how do these things converge and how things work after "Add", but with a failure at later stages.